### PR TITLE
[WIP] svg: support <circle> element

### PR DESCRIPTION
**Pull request by stshine**
Add basic support for SVG <circle> element.

 Currently, this only support filling with solid color.

---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17681)

### CODE REVIEW

**Comment by stshine** [comment]
This depends on #17649 get merged and https://github.com/servo/webrender/pull/1461 get resolved first, but IMO it can accepts reviews.

File: components/script/dom/svgelement.rs [hunk 1]
```diff
@@ -2,26 +2,83 @@
...
+use style_traits::PARSING_MODE_ALLOW_UNITLESS_LENGTH;
+use style::properties::declaration_block::{Importance, PropertyDeclarationBlock};
+use style::properties::{parse_one_declaration_into, SourcePropertyDeclaration};
 use style::element_state::ElementState;
+use style::properties::PropertyId;
 
 #[dom_struct]
 pub struct SVGElement {
     element: Element,
+    presentation_attributes: DOMRefCell<PropertyDeclarationBlock>
```
Comment (emilio) [inline components/script/dom/svgelement.rs]: You probably want to use `Arc<Locked<PropertyDeclarationBlock>>`, and use the document write guard to modify it in place. That way passing it to style is just a refcount bump.

File: components/script/dom/svgelement.rs [hunk 2]
```diff
@@ -2,26 +2,83 @@
...
+
+pub trait LayoutSVGElementHelpers {
+    fn presentation_attributes(&self) -> PropertyDeclarationBlock;
+}
+
+impl LayoutSVGElementHelpers for LayoutJS<SVGElement> {
+    #[allow(unsafe_code)]
+    fn presentation_attributes(&self) -> PropertyDeclarationBlock {
+        unsafe {
+            (*self.unsafe_get()).presentation_attributes.borrow_for_layout().clone()
```
Comment (emilio) [inline components/script/dom/svgelement.rs]: This clone is quite expensive actually, see above for what may be better.

File: components/script/dom/svgelement.rs [hunk 3]
```diff
@@ -30,4 +87,15 @@ impl VirtualMethods for SVGElement {
     fn super_type(&self) -> Option<&VirtualMethods> {
         Some(self.upcast::<Element>() as &VirtualMethods)
     }
+
+    fn attribute_mutated(&self, attr: &Attr, mutation: AttributeMutation) {
+        self.super_type().unwrap().attribute_mutated(attr, mutation);
+
+        match attr.local_name() {
+            &local_name!("fill") => {
+                self.mutate_svg_prehint(attr);
```
Comment (emilio) [inline components/script/dom/svgelement.rs]: You mean `mutate_svg_preshint`?

File: components/script/dom/svgsvgelement.rs [hunk 4]
```diff
@@ -70,6 +85,30 @@ impl VirtualMethods for SVGSVGElement {
...
+    // When the <svg> element is appended to a tree and its parent is an html element,
+    // ensure a geometry key for it.
+    fn bind_to_tree(&self, tree_in_doc: bool) {
+        if self.upcast::<Node>().GetParentNode().unwrap().is::<HTMLElement>() {
+                if self.geometry_key.get().is_none() {
+                    let window = window_from_node(self);
+                    let key = window.image_cache().create_geometry_key();
+                    self.geometry_key.set(Some(key));
+                }
+            }
```
Comment (emilio) [inline components/script/dom/svgsvgelement.rs]: nit: Indentation is off.

File: components/style/properties/longhand/svg.mako.rs [hunk 5]
```diff
@@ -158,3 +158,27 @@ ${helpers.predefined_type("mask-image", "ImageLayer",
     animation_value_type="discrete",
     flags="CREATES_STACKING_CONTEXT",
     has_uncacheable_values="True" if product == "gecko" else "False")}
+
+${helpers.predefined_type(
```
Comment (emilio) [inline components/style/properties/longhand/svg.mako.rs]: Are these actual CSS properties? This will allow them to be parsed from style attributes, which may not be what you want.

File: components/style/properties/properties.mako.rs [hunk 6]
```diff
@@ -820,11 +821,14 @@ pub enum DeclaredValue<'a, T: 'a> {
 /// extra discriminant word) and synthesize dependent DeclaredValues for
 /// PropertyDeclaration instances as needed.
 #[derive(Clone, PartialEq, Eq, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
 pub enum DeclaredValueOwned<T> {
     /// A known specified value from the stylesheet.
     Value(T),
     /// An unparsed value that contains `var()` functions.
-    WithVariables(Arc<UnparsedValue>),
+    WithVariables(
```
Comment (emilio) [inline components/style/properties/properties.mako.rs]: nit: I think this would look nicer changing it to:

```rust
WithVariables {
    #[cfg_attr(feature = "servo", ...)]
    value: Arc<UnparsedValue>,
}
```

File: components/style/properties/properties.mako.rs [hunk 7]
```diff
@@ -1108,7 +1114,9 @@ pub enum PropertyDeclaration {
     /// A css-wide keyword.
     CSSWideKeyword(LonghandId, CSSWideKeyword),
     /// An unparsed value that contains `var()` functions.
-    WithVariables(LonghandId, Arc<UnparsedValue>),
+    WithVariables(LonghandId,
+                  #[cfg_attr(feature = "servo", ignore_heap_size_of = "Arc")]
```
Comment (emilio) [inline components/style/properties/properties.mako.rs]: ditto

File: components/style/servo/restyle_damage.rs [hunk 8]
```diff
@@ -218,7 +218,8 @@ fn compute_damage(old: &ServoComputedValues, new: &ServoComputedValues) -> Servo
         get_inheritedbox.direction, get_inheritedbox.writing_mode,
         get_text.text_decoration_line, get_text.unicode_bidi,
         get_inheritedtable.empty_cells, get_inheritedtable.caption_side,
-        get_column.column_width, get_column.column_count
+        get_column.column_width, get_column.column_count,
+        get_inheritedsvg.fill, get_svg.cx, get_svg.cy, get_svg.r
```
Comment (emilio) [inline components/style/servo/restyle_damage.rs]: Ok, so you actually want a style struct field, but _not_ exposing it as a CSS property... So at least you need to use `internal="True"` on the property definition, and document that it's not web-exposed.

Comment (emilio) [inline components/style/servo/restyle_damage.rs]: Also, are you sure you want changes to `fill` to fully reflow the element? Why is that needed?

**Review by emilio — CHANGES_REQUESTED** [review]
This is super-exciting @stshine! \o/

File: components/script/dom/svgelement.rs (hunk 3 above)
Comment (stshine) [inline components/script/dom/svgelement.rs]: Indeed :)

File: components/style/properties/longhand/svg.mako.rs (hunk 5 above)
Comment (stshine) [inline components/style/properties/longhand/svg.mako.rs]: Yes they are, according the SVG2 spec every SVG presentation attribute is a css property. For example, you can do this in Chrome.

File: components/style/properties/properties.mako.rs (hunk 6 above)
Comment (stshine) [inline components/style/properties/properties.mako.rs]: Will do.

File: components/style/servo/restyle_damage.rs (hunk 8 above)
Comment (stshine) [inline components/style/servo/restyle_damage.rs]: To be honest I am not really familiar with the incremental reflow infrastructure, will try to resolve this.

**Comment by jdm** [comment]
@stshine Apologies for the long delay in reviewing these changes. Do you plan to continue working on this if we find a reviewer?

**Comment by jdm** [comment]
Tracking this work in #12975 but closing due to lack of activity.
